metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -109,6 +110,192 @@ inline void unset_bit(uint64_t *v,
     }
 }
 
+inline uint64_t atomic_fetch(const sdsl::int_vector<> &vector, uint64_t i,
+                             std::mutex &backup_mutex,
+                             int memorder = __ATOMIC_SEQ_CST) {
+    size_t width = vector.width();
+    size_t bit_pos = i * width;
+    uint64_t mask = (1llu << width) - 1;
+    if (width + 7 > 64) {
+        // there is no way to reliably modify without a mutex
+        std::lock_guard<std::mutex> lock(backup_mutex);
+        return vector[i];
+    } else if ((bit_pos & 0x3F) + width <= 64) {
+        // element fits in an aligned word
+        const uint64_t *word = &vector.data()[bit_pos >> 6];
+        return (__atomic_load_n(word, memorder) >> (bit_pos & 0x3F)) & mask;
+#if defined(MODE_TI) && defined(__CX16__)
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        const __uint128_t *word = &reinterpret_cast<const __uint128_t*>(vector.data())[bit_pos >> 7];
+        return (__atomic_load_n(word, memorder) >> (bit_pos & 0x7F)) & mask;
+#endif
+    } else {
+        uint8_t shift = bit_pos & 0x7;
+        const uint8_t *word = &reinterpret_cast<const uint8_t*>(vector.data())[bit_pos >> 3];
+        if (shift + width <= 8) {
+            // read from a byte
+            return (__atomic_load_n(word, memorder) >> shift) & mask;
+        } else if (shift + width <= 16) {
+            // unaligned read from two bytes
+            return (__atomic_load_n((const uint16_t*)word, memorder) >> shift) & mask;
+        } else if (shift + width <= 32) {
+            // unaligned read from four bytes
+            return (__atomic_load_n((const uint32_t*)word, memorder) >> shift) & mask;
+        } else if (shift + width <= 64) {
+            // unaligned read from eight bytes
+            return (__atomic_load_n((const uint64_t*)word, memorder) >> shift) & mask;
+        } else {
+            assert(false);
+        }
+    }
+
+    return 0;
+}
+
+inline uint64_t atomic_fetch_and_add(sdsl::int_vector<> &vector, uint64_t i,
+                                     uint64_t count,
```

Comment:
what is count?

I had to look at the code to get this.

Call it 'delta', or 'x', or 'val', ...

Reply:
I'll call it val to be consistent with `atomic_exchange`